examples/eSealing/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+//!   witness (sender_account) was in possession of that file at that time.
+//!
+//! Only accounts can register a file hash. During the registration, the
+//! timestamp is recorded together with the witness (sender_account) that
+//! registered the file hash. Each file hash can only be registered once.
+//! This is because we want to record the first time when the witness
+//! (sender_account) had access to that file. Re-registering a file hash (by a
+//! different witness) would not prove that the second witness is also in
+//! possession of that file or that it just observed the file hash during the
+//! initial registration hence a file hash can only be registered once.
+use concordium_std::*;
+
+/// The different errors the contract can produce.
+#[derive(Serialize, Debug, PartialEq, Eq, Reject, SchemaType)]
+enum ContractError {
+    /// Failed parsing the parameter.
+    #[from(ParseError)]
+    ParseParams,
+    /// Failed logging: Log is full.
+    LogFull,
+    /// Failed logging: Log is malformed.
+    LogMalformed,
+    /// Only accounts can register a file hash.
+    OnlyAccount,
+    /// Each file hash can only be registered once.
+    AlreadyRegistered,
+}
+
+/// Mapping the logging errors to ContractError.
+impl From<LogError> for ContractError {
+    fn from(le: LogError) -> Self {
+        match le {
+            LogError::Full => Self::LogFull,
+            LogError::Malformed => Self::LogMalformed,
+        }
+    }
+}
+
+/// The state tracked for each file.
+#[derive(Serial, Deserial)]
+struct FileState {
+    /// The timestamp when this file hash was registered.
+    timestamp: Timestamp,
+    /// The witness (sender_account) that registered this file hash.
+    witness:   AccountAddress,
+}
+
+/// FileHash type is a fixed-length byte array.
+type FileHash = [u8; 32];
+
+/// The contract state.
+#[derive(Serial, DeserialWithState, StateClone)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    files: StateMap<FileHash, FileState, S>,
+}
+
+impl<S: HasStateApi> State<S> {
+    /// Create a new state with no files registered.
+    fn new(state_builder: &mut StateBuilder<S>) -> Self {
+        State {
+            files: state_builder.new_map(),
+        }
+    }
+
+    /// Check if a file exists.
+    fn file_exists(&self, file_hash: &FileHash) -> bool {
+        let file = self.files.get(file_hash);
+
+        file.is_some()
+    }
+
+    /// Get recorded timestamp and witness from a specific file hash.
+    fn get_timestamp_and_witness(
+        &self,
+        file_hash: FileHash,
+    ) -> ReceiveResult<(Timestamp, AccountAddress)> {
+        let record = self.files.get(&file_hash);
+
+        match record {
+            // File exists.
+            Some(entry) => Result::Ok((entry.timestamp, entry.witness)),
+            // File does not exist.
+            None => Result::Ok((Timestamp::from_timestamp_millis(0), AccountAddress([0u8; 32]))),
```

Comment:
I'm not sure why we are not returning `None` in this case. That is our usual approach at least. Is there a reason why we can't return an option here instead? These "default values" are not the Rust way :)